packages/pigeon/lib/java_generator.dart

```diff
@@ -508,30 +500,52 @@ class JavaGenerator extends StructuredGenerator<JavaOptions> {
-            } else {
-              indent.addScoped('{', '});', () {
-                const String output = 'output';
-                indent.writeln('@SuppressWarnings("ConstantConditions")');
-                if (func.returnType.baseName == 'int') {
+            indent.addScoped('{', '});', () {
+              indent.writeScoped('if (channelReply instanceof List) {', '} ',
+                  () {
+                indent.writeln(
+                    'List<Object> listReply = (List<Object>) channelReply;');
+                indent.writeScoped('if (listReply.size() > 1) {', '} ', () {
                   indent.writeln(
-                      '$returnType $output = channelReply == null ? null : ((Number) channelReply).longValue();');
-                } else if (isEnum(root, func.returnType)) {
-                  if (func.returnType.isNullable) {
+                      'result.error(new FlutterError((String) listReply.get(0), (String) listReply.get(1), (String) listReply.get(2)));');
+                }, addTrailingNewline: false);
+                if (!func.returnType.isNullable && !func.returnType.isVoid) {
+                  indent.addScoped('else if (listReply.get(0) == null) {', '} ',
+                      () {
                     indent.writeln(
-                        '$returnType $output = channelReply == null ? null : $returnType.values()[(int) channelReply];');
+                        'result.error(new FlutterError("null-error", "Flutter api returned null value for non-null return value.", ""));');
+                  }, addTrailingNewline: false);
+                }
+                indent.addScoped('else {', '}', () {
+                  if (func.returnType.isVoid) {
+                    indent.writeln('result.success(null);');
                   } else {
-                    indent.writeln(
-                        '$returnType $output = $returnType.values()[(int) channelReply];');
+                    const String output = 'output';
+                    indent.writeln('@SuppressWarnings("ConstantConditions")');
+                    if (func.returnType.baseName == 'int') {
+                      indent.writeln(
+                          '$returnType $output = listReply.get(0) == null ? null : ((Number) listReply.get(0)).longValue();');
+                    } else if (isEnum(root, func.returnType)) {
+                      if (func.returnType.isNullable) {
+                        indent.writeln(
+                            '$returnType $output = listReply.get(0) == null ? null : $returnType.values()[(int) listReply.get(0)];');
+                      } else {
+                        indent.writeln(
+                            '$returnType $output = $returnType.values()[(int) listReply.get(0)];');
+                      }
+                    } else {
+                      indent.writeln(
+                          '$returnType $output = ${_cast('listReply.get(0)', javaType: returnType)};');
+                    }
```

Comment:
It's hard to see/verify that every case here is writing a `$returnType $output =` expression here; I would recommend:

```
final String outputExpression;
// all the same logic here, but instead of `indent.writeln('$returnType $output = ...');`
// do `outputExpression = '...';
// then after all the if/else-if chains:
indent.writeln('$returnType $output = $outputExpression');
```

Then it's easier to see what is the same vs different about the output in each case, and it also won't compile if anyone ever accidentally introduces a path that doesn't set `outputExpression`.